internal/server/middleware.go

```diff
@@ -49,63 +50,80 @@ func DatabaseMiddleware(db *gorm.DB) gin.HandlerFunc {
+	uniqueID := c.Request.Header.Get("Infra-Destination")
+	if uniqueID == "" {
+		return nil
+	}
 
-			switch len(destinations) {
-			case 0:
-				// destination does not exist yet, noop
-			case 1:
-				destination := destinations[0]
-				// only save if there's significant difference between LastSeenAt and Now
-				if time.Since(destination.LastSeenAt) > time.Second {
-					destination.LastSeenAt = time.Now()
-					if err := access.SaveDestination(c, &destination); err != nil {
-						sendAPIError(c, err)
-						return
-					}
-				}
-			default:
-				sendAPIError(c, fmt.Errorf("multiple destinations found for unique ID %q", uniqueID))
-				return
+	// TODO: use GetDestination(ByUniqueID())
+	destinations, err := access.ListDestinations(c, uniqueID, "", &models.Pagination{})
+	if err != nil {
+		return err
+	}
+
+	switch len(destinations) {
+	case 0:
+		// destination does not exist yet, noop
+		return nil
+	case 1:
+		destination := destinations[0]
+		// only save if there's significant difference between LastSeenAt and Now
+		if time.Since(destination.LastSeenAt) > time.Second {
+			destination.LastSeenAt = time.Now()
+			if err := access.SaveDestination(c, &destination); err != nil {
+				return fmt.Errorf("failed to update destination lastSeenAt: %w", err)
 			}
 		}
+		return nil
+	default:
+		return fmt.Errorf("multiple destinations found for unique ID %q", uniqueID)
+	}
+}
 
-		c.Next()
+// TODO: remove duplicate in access package
+func getDB(c *gin.Context) *gorm.DB {
+	db, ok := c.MustGet("db").(*gorm.DB)
+	if !ok {
+		return nil
 	}
+
+	return db
 }
 
-// AuthenticationMiddleware validates the incoming token
-func AuthenticationMiddleware() gin.HandlerFunc {
+// authenticatedMiddleware is applied to all routes that require authentication.
+// It validates the access key, and updates the lastSeenAt of the user, and
+// possibly also of the destination.
+func authenticatedMiddleware() gin.HandlerFunc {
 	return func(c *gin.Context) {
-		if err := RequireAccessKey(c); err != nil {
+		db := getDB(c)
+		authnUser, err := requireAccessKey(db, c.Request)
+		if err != nil {
 			sendAPIError(c, err)
 			return
 		}
 
+		// TODO: save authnUser in a single key
+		c.Set("key", authnUser.AccessKey)
+		c.Set("identity", authnUser.User)
+
+		if err := updateDestinationLastSeenAt(c); err != nil {
+			sendAPIError(c, err)
+			return
+		}
 		c.Next()
 	}
 }
 
-// RequireAccessKey checks the bearer token is present and valid
-func RequireAccessKey(c *gin.Context) error {
-	val, ok := c.Get("db")
-	if !ok {
-		return errors.New("could not find db in context")
-	}
-
-	db, ok := val.(*gorm.DB)
-	if !ok {
-		return errors.New("unknown db type in context")
-	}
+type authenticatedUser struct {
+	AccessKey *models.AccessKey
+	User      *models.Identity
+}
 
-	header := c.Request.Header.Get("Authorization")
+// requireAccessKey checks the bearer token is present and valid
+func requireAccessKey(db *gorm.DB, req *http.Request) (authenticatedUser, error) {
```

Comment:
What happens in this function when the db is nil?

Reply:
I expect it would panic if the `db` is nil. From the users perspective that wouldn't be any different than what we have today (500 status code).